### ISSUE TITLE
Improper handling of retries of SPDM requests

### ISSUE THREAD

**Issue by jcahill0** [DMTF/libspdm#491]
DSP0274 allows retries of SPDM requests, but not SPDM responses.  The likely high level cause for an SPDM request to be retried is the requester to timeout just as the response comes in or some other delay prevents the response from being recognized.  These could include bridge delays on the transport's bus, marginal responder timing, or even the requester's transport software stack delaying the response propagation to the application.

Neither libspdm's requester library nor responder library have proper handling for a retry.

A retry from an application to the libspdm responder code will fail at the check against the state since it thinks the state incremented.  This pattern occurs throughout the SPDM responder code.  Below is an example for NEGOTIATE_ALGORITHMS.
```
return_status spdm_get_response_algorithms(IN void *context,
...
    if (spdm_context->connection_info.connection_state !=
        LIBSPDM_CONNECTION_STATE_AFTER_CAPABILITIES) {
        return libspdm_generate_error_response(spdm_context,
                         SPDM_ERROR_CODE_UNEXPECTED_REQUEST,
                         0, response_size, response);
    }
...
    spdm_set_connection_state(spdm_context,
                  LIBSPDM_CONNECTION_STATE_NEGOTIATED);

    return RETURN_SUCCESS;
}
```

Now a retry by the application to libspdm requester code for NEGOTIATE_ALGORITHMS will fail differently.  The transcripts between the requester and responder will mismatch.  This is really impacts both the requester and responder code.  The requester only records in the transcript when it gets a response in time.
```
    status = libspdm_append_message_a(spdm_context, &spdm_request,
                       spdm_request.length);
    if (RETURN_ERROR(status)) {
        return RETURN_SECURITY_VIOLATION;
    }

    status = libspdm_append_message_a(spdm_context, &spdm_response,
                       spdm_response_size);
    if (RETURN_ERROR(status)) {
        return RETURN_SECURITY_VIOLATION;
    }
``` 
While the responder code records the transcript when it sends a response, and doesn't try to conditionality interaction with the transcript upon receiving a retry.
```
    status = libspdm_append_message_a(spdm_context, spdm_request,
                       spdm_request_size);
    if (RETURN_ERROR(status)) {
        return libspdm_generate_error_response(spdm_context,
                         SPDM_ERROR_CODE_UNSPECIFIED, 0,
                         response_size, response);
    }

    status = libspdm_append_message_a(spdm_context, spdm_response,
                       *response_size);
    if (RETURN_ERROR(status)) {
        return libspdm_generate_error_response(spdm_context,
                         SPDM_ERROR_CODE_UNSPECIFIED, 0,
                         response_size, response);
    }

    spdm_set_connection_state(spdm_context,
                  LIBSPDM_CONNECTION_STATE_NEGOTIATED);

    return RETURN_SUCCESS;
}
```

I filed a GitHub issue against DSP0274 to clarify the relationship between a retry and and the transcript.  Pushing the transcript portion to another issue could be preferable because of the issue against DSP0274, and then a new issue could be linked against https://github.com/DMTF/Security-TF/issues/1745, the issue I filed earlier.

**Comment by jcahill0** [comment]
@bhenning10

**Comment by jyao1** [comment]
Agree this is an issue in libspdm, which has not consider retry, yet.

However, we need get the precise definition on "retry" in SPDM spec. I believe different people has different opinion.
E.g. Some people think retry must be identical, some people think retry could be with different parameter.

Also, an open question: how a responder distinguish a "retry" or "new" command, even if the command are same.
E.g. What happen if responder receive a request to get measurement index 3 twice? Technically, it could be a new command, besides a retry, because SPDM spec does not forbid you send get measurement with same index twice.

I think we need resolve it after we reach consensus.

**Comment by jcahill0** [comment]
@jyao1 , agreed that the Security-TF needs to first consolidate on a complete definition of retry from a new command and how to address the more unique commands like GET_MEASUREMENTS.  So we can use this issue as a reminder to address it.

**Comment by rw8896** [comment]
The retry behavior sort of depends on how the application would like to handle the lower layer failures in the application code.

libspdm generates SPDM commands/responses for application to transport SPDM messages via its lower layer stack (e.g. MCTP). If application code detects the lower layer failures, developers could add their own retry mechanisms (in addition to the lower layer's built-in retransmission) without regenerating the SPDM commands or responses.

But if it's just timeout with no other lower layer failures, (the message reaches the peer successfully), the requester application probably shouldn't retry the failed command but needs to send GET_VERSIONS to resync the connection between both ends instead.

**Comment by jcahill0** [comment]
@rw8896 ,yes, I agree the retry behavior likely depends on the application code; we seem to have different perspectives as to whether it should be the higher layer above libspdm or in the transport layer below libspdm.  I've written up the issue as though the application saw the timeout of an SPDM request traverse upward from the transport layer, and is now reissuing the request.

I'm going to use the MCTP transport as the basis for this issue, mostly because that is what I'm familiar with, but also because it has been a cause for robust error handling in higher level protocols.

MCTP as a transport doesn't have generic built in retries for non-Control MsgTypes.  Same for MCTP's bindings to other transports with very fundamental exceptions: e.g NAKs on Addr and Data, which mean the full packet was not received and therefore is not actually outstanding to the destination.  So while it is technically possible for the application's transport layer to libspdm to do the retries, the implementations are likely to wildly vary and the retries will be hidden from the higher level application, which is responsible for the high level policies,  using libspdm.

Because of the limitations of MCTP's reliability as a transport, MCTP's lack of guidelines, and DSP0274 specifically saying retries of SPDM requests are valid, I disagree with the statement that the application shouldn't retry the timed out command and should resync with GET_VERSIONS.  It appears valid at an SPDM spec level, is useful and robust for simple topologies.  Topologies with devices that use very large Certificate Chains is one situation where starting attestation all over is not particularly desired.

**Comment by rw8896** [comment]
@jcahill0 My take is that the retry behavior depends on the application and transport layer owned by libspdm user. 
Agree with you that GET_VERSIONS has its cost and is kind of a last resort solution for the timeout. To be more specific,
0) If the transport layer is not reliable, libspdm users could add their error handling for example, if mctp_send_msg(spdm_message) failed, users could retry mctp_send_msg(spdm_message). This doesn't involve libspdm.
1) If the timeout command doesn't change the state, yes, the user application could just retry it. E.g, a periodic CHALLENGE command timeout, the requester won't need GET_VERSIONS to resync the state.
2) If the timeout command will change the state, it might be better to send GET_VERSIONS to resync the state.

**Comment by jyao1** [comment]
1/10/2022 - wait for spec clarification.

**Comment by jcahill0** [comment]
Spec now clarifies that retries are identical in regards to SPDM payload

https://github.com/DMTF/Security-TF/pull/1778/files

**Comment by jyao1** [comment]
I think the flow is:
1) responder caches the last received message and last sent message.
2) responder compares the current message with last received message.
3) if current message == last received message /*retry*/, then responder sends the last sent message again.
4) else /*not retry*/, then responder processes current message normally.

NOTE: That will increase the context size (or scratch buffer) to hold the last received message and last sent message.

**Comment by rw8896** [comment]
@jyao1 I wonder if retry has to be done within libspdm. 
Integrators seem to be able to implement these steps without any libspdm support, right?

**Comment by jyao1** [comment]
Currently, libspdm does not support "retry".

We know a "retry" message must be identical to previous.
However, we don't know if an identical message is retry or not. A) It may be a retry. B) It may be a new message, which happens to be identical.

The problem is transcript calculation. For example, measurement transcript.
According to SPDM "Because a retried message is identical to the first, a retried message shall not be used in transcript hash calculations.". But how could the receiver knows it is retry or an identical new message?

https://github.com/DMTF/SPDM-WG/issues/1763 will address that.
